src/fundus/scraping/session.py

```diff
@@ -0,0 +1,93 @@
+from typing import Optional
+
+import requests.adapters
+
+from fundus.logging import basic_logger
+
+_default_header = {"user-agent": "Fundus"}
+
+
+class SessionHandler:
+    """Object for handling  project global request.Session
+
+    The session life cycle consists of three steps which can be repeated indefinitely:
+    Build, Supply, Teardown.
+    Initially there is no session build within the session handler. When a session is requested
+    with get_session() either a new one is created with _session_factory() or the session handler's
+    existing one returned. Every subsequent call to get_session() will return the same
+    response.Session object. If close_current_session() is called, the current session will be
+    tear-downed and the next call to get_session() will build a new session.
+    """
+
+    def __init__(self, pool_connections: int = 50, pool_maxsize: int = 50):
+        self.session: Optional[requests.Session] = None
+        self.pool_connections = pool_connections
+        self.pool_maxsize = pool_maxsize
+
+    def _session_factory(self) -> requests.Session:
+        """Builds a new Session
+
+        This returns a new client session build from pre-defined configurations:
+        - pool_connections: 50
+        - pool_maxsize: 50
+        - hooks = {'response': raise_for_status(), _response_log():}
+
+        Returns:
+            A new requests.Session
+        """
+
+        basic_logger.debug("Creating new session")
+        session = requests.Session()
+
+        def _response_log(response: requests.Response, *args, **kwargs) -> None:
+            history = response.history
+            previous_status_codes = [f"({response.status_code})" for response in history] if history else []
+            status_code_chain = " -> ".join(previous_status_codes + [f"({response.status_code})"])
+            basic_logger.debug(
+                f"{status_code_chain} <{response.request.method} {response.url!r}> "
+                f"took {response.elapsed.total_seconds()} second(s)"
+            )
+
+        # hooks
+        response_hooks = [lambda response, *args, **kwargs: response.raise_for_status(), _response_log]
+        session.hooks["response"].extend(response_hooks)
+
+        # adapters
+        session.mount(
+            "http://",
+            requests.adapters.HTTPAdapter(pool_connections=self.pool_connections, pool_maxsize=self.pool_maxsize),
+        )
+        session.mount(
+            "https://",
+            requests.adapters.HTTPAdapter(pool_connections=self.pool_connections, pool_maxsize=self.pool_maxsize),
+        )
```

Comment:
I made some changes so that `pool_connections` == `len(publishers)` and `pool_maxsize` == 0.

Reply:
- Yes, the connection pool caches established connections per host, and with every run, we utilize len(publishers) hosts. In this context, "Host" covers subdomains as well.
- Yes, the 0 is a type. It's supposed to be a 1 because there is only one thread per host active and we never encounter a situation where two threads want to connect to the same host.

Reply:
Very nice :)